### DISCUSSION TITLE
How do i make the game handle high amount of sprites?

### DISCUSSION THREAD

**Discussion by Snirozu** [HaxeFlixel/flixel#2579] [Q&A]
So i'm trying to make the game draw 400 sprites (with every single sprite having different graphic) but when i do that the framerate drops by about 70, for some reason this doesn't happen when i change the graphic of the sprites to be the same or when i change the target to be HTML5.

**Comment by Geokureli** [comment]
what fps are you aiming for?

**Reply by Snirozu** [reply]
144

**Reply by Geokureli** [reply]
I honestly didn't know it goes that high. most monitors don't even refresh that fast, why not use 60fps and set `FlxG.fixedTimeStep = false;`
https://api.haxeflixel.com/flixel/FlxG.html#fixedTimestep

95% of users won't notice a difference

**Reply by Snirozu** [reply]
works kinda better now, thanks

**Reply by Geokureli** [reply]
what fps are you getting now?

**Reply by Snirozu** [reply]
about 70

**Reply by RichardBray** [reply]
144!! Wow, I think there's a Linus tech tips video saying that over a certain amount doesn't make a bit difference, maybe 120 or something, but for most people and monitors 60 should be more than enough.

**Comment by Geokureli** [comment]
you should create a copy of the [FlxBunnyMark Demo](https://haxeflixel.com/demos/FlxBunnymark/) and try out different numbers of objects to get a sense for your limitations. you can see how many colliding sprites you can handle, or add other features in to see how intensive they are.

On Html5, at 60fps i can get 7000 sprites moving around before i get issues. and about 1600 sprites colliding with each other before the frame rate dips. Also note how you can toggle the fixed timestep to make it feel smoother when the framerate dips a tiny bit

**Reply by Snirozu** [reply]
well the demo has stable 144 fps but when i change 26 line in bunny class to
``` haxe
loadGraphic('assets/wabbit_alpha-${new FlxRandom().int(1, 3)}.png');
```
it drops to like 40 fps while the game has 1000 sprites and 80fps while 400 sprites
and yeah it doesn't happen on html5 because it smoothes the game for some reason

**Reply by Geokureli** [reply]
can you share the sprites you're using? based on the name I assume they have pixels with varying alphas? if possible I would avoid large amounts of sprites whose pixels are anything other than 0 or 1

**Reply by Geokureli** [reply]
also new `new FlxRandom().int(1, 3)` should be `FlxG.random.int(1, 3)`. Flixel makes an instance for you, making a new instance of random 1000s of times is gonna waste a ton of memory.

I've updated the docs so future people know
https://github.com/HaxeFlixel/flixel/pull/2583/files

**Reply by Snirozu** [reply]
[assets.zip](https://github.com/HaxeFlixel/flixel/files/8912477/assets.zip)
they're just recolors of the demo bunny sprite

**Reply by Geokureli** [reply]
You're right, it seems to be able to draw the same graphic many times in a row in a more performant way than with many different graphics. with your new assets I  can only get about 1000 bunnies to run at 60fps, however I was able to get up to 6000 by sorting the _bunnies group by graphic. I added this to the bottom of changeBunnyNumber
```hx
if (add)
{
    _bunnies.members.sort((a, b) ->
    {
        return Std.parseInt(b.graphic.key.split("-").pop()) - Std.parseInt(a.graphic.key.split("-").pop());
    });
}
```

This causes all the yellow bunnies to be on top, green below them and purple on bottom. if you're game doesn't have any complex layering you could leverage this for better performance.

typically in my games I have a group for each type of sprite and I add them all to the state and it ends up drawing them by groups by default, for instance:

```hx
enemyGroup = new FlxTypedGroup<Enemy>();
playerBulletGroup = new FlxTypedGroup<PlayerBullet>();
enemyBulletGroup = new FlxTypedGroup<EnemyBullet>();
playerGroup = new FlxTypedGroup<Player>();
add(enemyGroup);
add(playerBulletGroup);
add(enemyBulletGroup);
add(playerGroup);
```
It's already a good idea to plan out your game's layering, to show higher priority things on top, but apparently it has performance benefits too! Hope this helps
